lib/Service/ImageService.php

```diff
@@ -0,0 +1,676 @@
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	public function uploadImagePublic(?int $documentId, string $newFileName, string $newFileContent, string $shareToken): array {
+		if (!$this->hasUpdatePermissions($shareToken)) {
+			throw new Exception('No write permissions');
+		}
+		$textFile = $this->getTextFilePublic($documentId, $shareToken);
+		$saveDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+		if ($saveDir !== null) {
+			$fileName = (string) time() . '-' . $newFileName;
+			$savedFile = $saveDir->newFile($fileName, $newFileContent);
+			return [
+				'name' => $fileName,
+				'id' => $savedFile->getId(),
+				'documentId' => $textFile->getId(),
+			];
+		} else {
+			return [
+				'error' => 'Impossible to get attachment directory',
+			];
+		}
+	}
+
+	/**
+	 * Copy a file from a user's storage in the attachment folder
+	 *
+	 * @param int $documentId
+	 * @param string $path
+	 * @param string $userId
+	 * @return array
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 * @throws \OCP\Lock\LockedException
+	 */
+	public function insertImageFile(int $documentId, string $path, string $userId): array {
+		$textFile = $this->getTextFile($documentId, $userId);
+		if (!$textFile->isUpdateable()) {
+			throw new Exception('No write permissions');
+		}
+		$imageFile = $this->getFileFromPath($path, $userId);
+		$saveDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+		if ($saveDir !== null) {
+			return $this->copyImageFile($imageFile, $saveDir, $textFile);
+		} else {
+			return [
+				'error' => 'Impossible to get attachment directory',
+			];
+		}
+	}
+
+	/**
+	 * @param File $imageFile
+	 * @param Folder $saveDir
+	 * @param File $textFile
+	 * @return array
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 */
+	private function copyImageFile(File $imageFile, Folder $saveDir, File $textFile): array {
+		$mimeType = $imageFile->getMimeType();
+		if (in_array($mimeType, ImageController::IMAGE_MIME_TYPES)) {
+			$fileName = (string) time() . '-' . $imageFile->getName();
+			$targetPath = $saveDir->getPath() . '/' . $fileName;
+			$targetFile = $imageFile->copy($targetPath);
+			// get file type and name
+			return [
+				'name' => $fileName,
+				'id' => $targetFile->getId(),
+				'documentId' => $textFile->getId(),
+			];
+		}
+		return [
+			'error' => 'Unsupported file type',
+		];
+	}
+
+	/**
+	 * Download and save an image from a link in the attachment folder
+	 *
+	 * @param int $documentId
+	 * @param string $link
+	 * @param string $userId
+	 * @return array
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	public function insertImageLink(int $documentId, string $link, string $userId): array {
+		$textFile = $this->getTextFile($documentId, $userId);
+		if (!$textFile->isUpdateable()) {
```

Comment:
Same as above: either check to prevent null pointer exception, or throw in `$this->getTextFile()`.